ADBDEV-6156 Count startup memory of each process when using resource groups

dnskvlnk opened the pull request:
Count the startup memory of each active process when using resource groups

Make the resource manager track the startup memory of each active backend so
that the runaway detector would estimate memory more accurately.

The startup memory is the memory that the backend consumes after startup before
the memory managers (Vmem tracker and resource groups) are initialized. The Vmem
tracker counts this memory as consumed by the segment, but after the backend was
assigned a resource group, this memory was not counted as consumed by the group.

This patch adds startup memory consumption to self->memUsage to make resource
groups consider this memory. 
Additionally, this patch slightly modifies the resGroupPalloc function so that
it takes startup memory into account. This is necessary to avoid changing or
complicating the logic of existing tests.

It is worth noting that this patch fixes the accounting of memory consumption by
only active backends (which execute the query). Accounting for the memory
occupied by idle backends is a more complex task that should be done separately.

Commit:
Count startup memory of each process when using resource groups

The goal of the patch is to make resource groups manager to track startup
memory of each process so that the redzone handler would estimate memory more
accurately, but the context needs to be brought.

Startup memory is considered correctly by the VMEM tracker, but it's not
accounted by the resource groups manager because on a process' startup this
memory is not added to self->memUsage. This happened because the startup memory
is considered on VMEM tracker initialization, but the ways VMEM tracker and
resource groups count memory are not consistent with each other.

This patch adds startup memory consumption to self->memUsage to make resource
groups consider this memory.

However, one should note that once a process is detached from a resource group
(it finished query execution), resource groups memory manager doesn't see this
memory again. This happens due to resource groups' design. We could've moved
inactive processes' memory to freeChunks, but that would drain freeChunks pretty
fast and cause new processes to fail because any process starts without a
resource group and acquires it later, thus it must add it's startup memory to
freeChunks at first. If freeChunks is already exhausted, a new process fails to
start. Also, there are other complexities in making resource groups manager to
fully consider startup memory. Thankfully, we have the VMEM tracker which
calculates it correctly, so if the resource groups manager fails to track it
correctly, we can still rely on it. It's impossible to solve this problem in
this patch because it's a design flaw rather than a bug.

BenderArenadata commented:
Failed job Deploy multiarch Dockerimages: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1807317

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/78383

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1807327

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1807328

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/78465

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1817041

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1817039

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1817048

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1818658

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1818659

RekGRpth commented:
Can you write some tests to check?

RekGRpth commented on src/include/utils/resgroup.h:
```diff
@@ -235,6 +235,9 @@ extern Oid ResGroupGetGroupIdBySessionId(int sessionId);
 extern char *getCpuSetByRole(const char *cpuset);
 extern void checkCpuSetByRole(const char *cpuset);
 
+extern void selfAddStartupChunks(int32 chunks);
+extern void selfSubStartupChunks(int32 chunks);
```
I think we should rename public functions to something like ResGroupAddStartupChunks and ResGroupSubStartupChunks or use one function ResGroupAppendStartupChunks with positive and negative arguments

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1850837

RekGRpth commented on src/test/isolation2/isolation2_resgroup_schedule:
```diff
@@ -60,3 +60,5 @@ test: resgroup/resgroup_dumpinfo
 
 test: resgroup/disable_resgroup
+
+test: resgroup/resgroup_startup_memory
```
Why did you add your new test after disabling resource groups?

dnskvlnk replied:
I moved it higher

RekGRpth commented on src/test/isolation2/output/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,47 @@
+DROP
+
+-- start_ignore
+! gpstop -rai;
```
may be add
```sql
! gpconfig -r gp_resource_manager;
```
before this line?

dnskvlnk replied:
This is done in `disable_resgroup` test

RekGRpth commented:
`resgroup/enable_resgroup` test hangs after applying patch.

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1880757

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1890216

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/79897

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1891649

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80245

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1911167

RekGRpth commented on src/test/regress/regress_gp.c:
```diff
@@ -637,12 +673,18 @@ resGroupPalloc(PG_FUNCTION_ARGS)
 
 	ResGroupGetMemInfo(&memLimit, &slotQuota, &sharedQuota);
-	size = ceilf(memLimit * ratio);
+	if (!startupConsidered)
+		size = ceilf(memLimit * ratio) - 12;
+		startupConsidered = true;
+	else {
```
seriously?!

dnskvlnk replied:
The ticket is not assigned to you and is not purposed for review

BenderArenadata commented:
Failed job Build ubuntu22 for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1924163

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1924162

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80289

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1926814

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80301

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1948493

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80365

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1961263

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80450

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80507

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80516

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1969490

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80543

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1970497

RekGRpth commented on src/test/isolation2/output/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,148 @@
+ group               
+(1 row)
+
+CREATE OR REPLACE FUNCTION resGroupPalloc(float) RETURNS int AS '/home/kds/gpdb/src/test/isolation2/../regress/regress.so', 'resGroupPalloc' LANGUAGE C READS SQL DATA;
```
```suggestion
CREATE OR REPLACE FUNCTION resGroupPalloc(float) RETURNS int AS '@abs_builddir@/../regress/regress@DLSUFFIX@', 'resGroupPalloc' LANGUAGE C READS SQL DATA;
```

dnskvlnk replied:
Done

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80563

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80606

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1974345

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80619

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80647

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1976376

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80660

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1977140

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/80692

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1979177

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/83439

Commit:
Try to fix diff caused by gpstop

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/83582

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/83693

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-6156

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/83850

bandetto commented on src/test/regress/regress_gp.c:
```diff
@@ -642,7 +646,12 @@ resGroupPalloc(PG_FUNCTION_ARGS)
 
 	ResGroupGetMemInfo(&memLimit, &slotQuota, &sharedQuota);
-	size = ceilf(memLimit * ratio);
+	if (!startupConsidered) {
+		size = ceilf(memLimit * ratio) - 12;
+		startupConsidered = true;
+	} else
+		size = ceilf(memLimit * ratio);
+
```
Maybe use allman brace style and move simpler case first?
```suggestion
	if (startupConsidered)
		size = ceilf(memLimit * ratio);
	else 
	{
		size = ceilf(memLimit * ratio) - 12;
		startupConsidered = true;
	}
```

dnskvlnk replied:
Yes, the style doesn't match the codebase. But I would leave the order to make it more visible that the first allocation acts differently

dnskvlnk replied:
I fixed the formatting

bandetto commented on src/test/regress/regress_gp.c:
```diff
@@ -642,7 +646,12 @@ resGroupPalloc(PG_FUNCTION_ARGS)
 	ResGroupGetMemInfo(&memLimit, &slotQuota, &sharedQuota);
-	size = ceilf(memLimit * ratio);
+	if (!startupConsidered) {
+		size = ceilf(memLimit * ratio) - 12;
```
If 12 chunks are always required for the startup memory, maybe it should be factored out as a macro or a constant?

dnskvlnk replied:
No, it's just a mistake, there should've been a function that retrieves the actual number of startup chunks, but it happened to be 12 on my configuration. I fixed this

bandetto commented on src/test/isolation2/output/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,79 @@
+-- Start a session which will be detached from a group when the query is done
+-- resource groups can't see startup chunks occupied by a detached session but
+-- the vmem tracker can
+0: SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE pid != (SELECT pg_backend_pid());
```
Why is this query required? What does exactly this test check?

dnskvlnk replied:
This test checks that the vmem tracker can see startup chunks of those processes which are not attached to any resource group. It starts the 0 session which terminates all other processes which can interfere this test's results and then creates a table just to make sure that this session is spawned on all segments. This session is then detached from a resource group and is tracked by vmem only.
Other sessions start transactions so that they can be attached to their resource groups. They allocate all available memory to a resource group. The last allocation should fail because of the vmem tracker, so we could make sure that the vmem tracker actually tracks the memory of those sessions which are not attached to a resource group

bandetto replied:
What was the expected behavior of this test before the patch?

bandetto replied:
Bump .__.

dnskvlnk replied:
Actually, you are right here, this test is redundant. It supposed to be sort of sanity check, but since it behaves just like it would behave without the patch, it doesn't make sense, so I removed it

bandetto commented:
Maybe add `sql/resgroup/resgroup_startup_memory.sql` and `expected/resgroup/resgroup_startup_memory.out` to `src/test/isolation2/.gitignore`? For some reason, not all the generated sql and expected files from `input/resgroup/` are ignored, but all tests inside `input/` folder are.

dnskvlnk commented:
@bandetto 
> Maybe add sql/resgroup/resgroup_startup_memory.sql and expected/resgroup/resgroup_startup_memory.out to src/test/isolation2/.gitignore? For some reason, not all the generated sql and expected files from input/resgroup/ are ignored, but all tests inside input/ folder are.

Yes, this is pretty strange, so I would agree with you. However, I'm not sure about ignoring .sql since they are actually tracked, but some are not, as far as I remember

Commit:
Use the actual number of startup chunks instaed of a heardcoded value

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-6156

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/83966

bandetto commented:
> However, I'm not sure about ignoring .sql since they are actually tracked, but some are not, as far as I remember

Tests which have input files in `input/` instead of `sql/` get their `sql/..sql` files generated out of files from the `input/` directory. It means that any `sql/..sql` file from a test that uses `input/` directory can safely be ignored. Seems that this ignoring thing is inconsistent, so it's is up to you to fix it in this patch :^)

dnskvlnk commented:
@bandetto I know that, but there are some tests which are generated from .source file but the respective .sql is still tracked. Though I agree with you that if a test  is generated then .sql should be ignored

Commit:
Add resGroupPallocIgnoreStartup which copies old Add resGroupPalloc

In order to run tests, we need the old version of resGroupPalloc, however, to
keep old tests intact, we need to add it under a new name, because the old
logic of resGroupPalloc isn't compatible with the new behaviour

Commit:
Add new test case which tests redzone

Commit:
Ignore generated file

Commit:
Ignore function creation output

Commit:
Add a comment for resGroupPallocIgnoreStartup

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/84568

Commit:
Consider startup memory only if allocation was successful

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/84658

bandetto commented on src/test/regress/regress_gp.c:
```diff
@@ -628,6 +631,39 @@ repeatPalloc(PG_FUNCTION_ARGS)
 }
 
+/*
+ * This function is a copy of resGroupPalloc that existed before the patch.
```
What patch? :D
Can you be a little more specific for future readers?

dnskvlnk replied:
Done

bandetto commented on src/test/regress/regress_gp.c:
```diff
@@ -628,6 +631,39 @@ repeatPalloc(PG_FUNCTION_ARGS)
+ * to maintain compatibility with the old tests which don't expect resGroupPalloc
+ * to allocate extra memory which is startupChunks
+ */
+PG_FUNCTION_INFO_V1(resGroupPallocIgnoreStartup);
+Datum
+resGroupPallocIgnoreStartup(PG_FUNCTION_ARGS)
+{
+	float ratio = PG_GETARG_FLOAT8(0);
+	int memLimit, slotQuota, sharedQuota;
+	int size;
+	int count;
+	int i;
+
+	if (!IsResGroupEnabled())
+		PG_RETURN_INT32(0);
+
+	ResGroupGetMemInfo(&memLimit, &slotQuota, &sharedQuota);
+	size = ceilf(memLimit * ratio);
+
+	count = size / 512;
+	for (i = 0; i < count; i++)
+		MemoryContextAlloc(TopMemoryContext, 512 * 1024 * 1024);
+
+	size %= 512;
+	MemoryContextAlloc(TopMemoryContext, size * 1024 * 1024);
+
+	PG_RETURN_INT32(0);
+}
```
Can both `resGroupPalloc...()` functions be DRYed so that they use one same function inside, maybe passing a boolean value to check if it should consider startup memory? I think it's unwanted to disperse the logic.

dnskvlnk replied:
No. Changing its signature would require rewriting other tests. I tried to avoid it
https://github.com/arenadata/gpdb/blob/ef84cdb85fbf05581bcbddc140c814b34487db7e/src/test/regress/regress_gp.c#L634-L639

bandetto replied:
I meant doing something like:
```c
PG_FUNCTION_INFO_V1(resGroupPalloc);
Datum
resGroupPalloc(PG_FUNCTION_ARGS)
{
	resGroupPallocImpl(..., false);
}

PG_FUNCTION_INFO_V1(resGroupPallocIgnoreStartup);
Datum
resGroupPallocIgnoreStartup(PG_FUNCTION_ARGS)
{
	resGroupPallocImpl(..., true);
}

static ...
resGroupPallocImpl(..., bool ignore_startup)
{
	...
}
```

dnskvlnk replied:
Done

bandetto commented on src/test/isolation2/output/resgroup/resgroup_bypass.source:
```diff
@@ -161,7 +161,7 @@ SELECT * FROM memory_result;
 ----------------+----------+---------
  rg_bypass_test | 0        | 0.0     
- rg_bypass_test | 1        | 12.0    
+ rg_bypass_test | 1        | 24.0    
```
Do we need to change these values, or we could use `resGroupPallocIgnoreStartup()` here? Or for lesser diff, leave the old name intact, and use a new name for the new allocation function?

dnskvlnk replied:
Why would we need this patch then? This diff shows that the startup memory is counted by all processes

bandetto commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,75 @@
+5q:
+
+-- This query will be killed by redzone because resGroupPallocIgnoreStartup just allocates
+-- the memory size of all resgroup memory. Before the patch, a group had no memory tracked
```
What patch?

dnskvlnk replied:
Done

bandetto reviewed (commented):
> However, one should note that once a process is detached from a resource group (it finished query execution), resource groups memory manager doesn't see this memory again.

Can you please elaborate what this means?

> We could've moved inactive processes' memory to freeChunks, but that would drain freeChunks pretty fast

Why?

Please explain these points, and why the added tests are needed in PR description (and maybe you can skip explaining some of parts in the description, if they are not really related to the patch, and shorten it to only describe key changes).

RekGRpth commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,75 @@
+-- Start a session which will be detached from a group when the query is done
+-- resource groups can't see startup chunks occupied by a detached session but
+-- the vmem tracker can
+0: SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE pid != (SELECT pg_backend_pid());
```
here and below
```suggestion
0: SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE pid != pg_backend_pid();
```
Do we really want to kill ALL backends except the current one?!

dnskvlnk replied:
Yes, that's the point of the test, I don't want other processes to interfere. I can change it if you explain what problems it can cause

Commit:
Bring out allocation logic to a dedicated function

Commit:
Reword comments so they would be more clear

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-6156

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/85888

Commit:
Remove redundant test

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/85900

bandetto commented on src/test/isolation2/output/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,38 @@
+! gpconfig -c runaway_detector_activation_percent -v 20;
+! gpstop -rai;
+
+CREATE OR REPLACE FUNCTION resGroupPalloc(float) RETURNS int AS '/home/kds/gpdb/src/test/isolation2/../regress/regress.so', 'resGroupPalloc' LANGUAGE C READS SQL DATA;
```
Path remained the same (without generic variable), but the tests have passed?

dnskvlnk replied:
Yes, I did it intentionally. Compilation path is defined during configuration stage, so it differs on different machines. Having such statement outside `ignore` block will cause this test to fail if your compilation path is not the same compared to the default path on ci. If this statement fails to complete, it's not a problem either, because those statements which use this function will indicate the problem

bandetto replied:
What's about using `@abs_builddir@` and `@DLSUFFIX@` used in input, in output as well? It's replaced in runtime by `pg_regress`. The paths don't have to be inside ignore block that way.

dnskvlnk replied:
Done

bandetto commented on src/test/regress/regress_gp.c:
```diff
@@ -628,28 +631,57 @@ repeatPalloc(PG_FUNCTION_ARGS)
-{
-	float ratio = PG_GETARG_FLOAT8(0);
+static bool startupConsidered = false;
+void
```
```suggestion
static void
```

Commit:
Make resGroupPallocImpl static

Co-authored-by: Vladimir Sarmin <v.sarmin@arenadata.io>

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/85984

Commit:
Fix library path for test output

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-6156

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/86650

KnightMurloc committed:
new test

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/90016

KnightMurloc committed:
fix test

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/90168

andr-sokolov commented on src/test/isolation2/expected/resgroup/resgroup_startup_memory.out:
```diff
@@ -0,0 +1,60 @@
+-- The runaway detector test. A query with a large number of slices should
+-- be terminated due to high memory consumption.
+select count(*) from t1 a1 join t1 a2 using(a) join t1 a3 using(a) join t1 a4 using(a) join t1 a5 using(a) join t1 a6 using(a) join t1 a7 using(a) join t1 a8 using(a) join t1 a9 using(a) join t1 a10 using(a);
+ERROR:  Canceling query because of high VMEM usage. current group id is 712716, group memory usage 133 MB, group shared memory quota is 102 MB, slot memory quota is 0 MB, global freechunks memory is 277 MB, global safe memory threshold is 277 MB (runaway_cleaner.c:197)  (seg1 slice10 172.18.0.3:6003 pid=88018) (runaway_cleaner.c:197)
```
Rewrite please the test so that the error about memory consumption before the patch and after the patch occurs for different reasons

KnightMurloc committed:
rework test

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/90809

andr-sokolov commented on src/test/regress/regress_gp.c:
```diff
@@ -641,8 +643,24 @@ resGroupPalloc(PG_FUNCTION_ARGS)
+	// At startup, the backend process is already consuming some amount of
+	// memory. In order not to complicate the logic of the tests, we take this
+	// memory into account when allocating memory for tests.
+	if (startUpMBToAccount)
```
Why do we need this check?

KnightMurloc replied:
The purpose of this function is to allocate N% of the memory from the size of the current resource group. Because at the time the function is called, the backend is already consuming at least startup memory. Therefore, this function reduces memory allocation by the size of startup memory.

KnightMurloc committed:
rework resGroupPalloc

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/90926

andr-sokolov commented on src/test/regress/regress_gp.c:
```diff
@@ -641,8 +643,23 @@ resGroupPalloc(PG_FUNCTION_ARGS)
+	// At startup, the backend process is already consuming some amount of
+	// memory. In order not to complicate the logic of the tests, we take this
+	// memory into account when allocating memory for tests.
+	if (startUpMbRemains > 0)
+	{
+		size = Max(0, size - startUpMbRemains);
+		startUpMbRemains = Max(0, startUpMbRemains - size);
+	}
```
The current code calls MemoryContextAlloc when size is 0.
When the input is startUpMbRemains = 12 and size = 4, the output is size = max(0, 4-12) = 0 and startUpMbRemains = Max(0,12-0) = 12. But the output startUpMbRemains should be 8
```suggestion
	if (startUpMbRemains >= size)
	{
		startUpMbRemains -= size;
		PG_RETURN_INT32(0);
	}

	size -= startUpMbRemains;
	startUpMbRemains = 0;
```

KnightMurloc replied:
done

andr-sokolov commented:
"but after the backend is assigned a resource group, this memory is not counted as consumed by the group." - it has been fixed, so let's write in past tense

andr-sokolov commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -1594,6 +1594,30 @@ selfDetachResGroup(ResGroupData *group, ResGroupSlotData *slot)
+ * startup memory consumpion, but let it be just for symmetry.
+ */
+void
+ResGroupProcSubStartupChunks(int32 chunks)
```
What about removing this function and calling `ResGroupProcAddStartupChunks(-startupChunks)` instead of `ResGroupProcSubStartupChunks(startupChunks)` or removing the chunks argument and using VmemTracker_GetStartupChunks()?

KnightMurloc replied:
removed.

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_move_query.source:
```diff
@@ -11,6 +11,10 @@
 -- end_matchsubs
 
+-- start_ignore
+! gpstop -rai;
```
Why this line is added?

KnightMurloc replied:
reverted

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_move_query.source:
```diff
@@ -135,7 +139,7 @@ SELECT num_running FROM gp_toolkit.gp_resgroup_status WHERE rsgname='rg_move_que
 2: SET ROLE role_move_query_mem_small;
 2: BEGIN;
-2: SELECT hold_memory_by_percent_on_qe(1,0.1);
+2: SELECT hold_memory_by_percent_on_qe(1,0.2);
```
Why 0.1 is replaced with 0.2 here and below?

KnightMurloc replied:
reverted

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,108 @@
+ALTER RESOURCE GROUP admin_group SET memory_shared_quota 0;
+ALTER RESOURCE GROUP default_group SET memory_shared_quota 0;
+
+create resource group rg1 with (cpu_rate_limit=20, memory_limit=15, memory_shared_quota=100, memory_spill_ratio=0);
```
Do we really need memory_spill_ratio?

KnightMurloc replied:
removed

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,108 @@
+ALTER RESOURCE GROUP admin_group SET memory_limit 0;
+ALTER RESOURCE GROUP default_group SET memory_limit 0;
+ALTER RESOURCE GROUP admin_group SET memory_shared_quota 0;
+ALTER RESOURCE GROUP default_group SET memory_shared_quota 0;
```
Why do we need to change memory_shared_quota?

KnightMurloc replied:
removed

KnightMurloc committed:
improve the test

KnightMurloc committed:
improve the test

KnightMurloc committed:
Merge branch 'adb-6.x-dev' into ADBDEV-6156

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/91031

KnightMurloc committed:
reduce diff

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/91084

KnightMurloc committed:
make test more stable and make sql in the test lowercase.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/92112

andr-sokolov commented on src/backend/utils/mmgr/vmem_tracker.c:
```diff
@@ -692,6 +700,8 @@ VmemTracker_UnregisterStartupMemory(void)
 		pg_atomic_sub_fetch_u32((pg_atomic_uint32 *) &MySessionState->sessionVmem,
 								startupChunks);
 
+	ResGroupProcAddStartupChunks(-startupChunks);
```
Why do we need to call ResGroupProcAddStartupChunks here? What happens without this call?

KnightMurloc replied:
removed

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,105 @@
+drop role if exists test2;
+drop resource group rg1;
+drop resource group rg2;
+drop resource group rg3;
```
do we use rg3?

KnightMurloc replied:
removed

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,105 @@
+'@abs_builddir@/../regress/regress@DLSUFFIX@', 'repeatPalloc'
+    language C reads sql data;
+
+create or replace function hold_memory(int) returns int as $$
+select * from repeatPalloc(1, $1)
+$$ language sql;
```
can we replace `hold_memory` with `repeatPalloc(40, 1)`?

KnightMurloc replied:
done

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,105 @@
+alter resource group rg2 set memory_limit 75;
+-- Test that the starting memory is visible to the resource group.
+1: set role test1;
+1&: select count(*) from t1 where pg_sleep(1) is not null;
```
can we use inject fault instead of pg_sleep?

KnightMurloc replied:
done

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,105 @@
+-- to avoid executing hold_memory before all backends are up.
+select gp_inject_fault_infinite('executor_pre_tuple_processed', 'suspend', dbid)
+    from gp_segment_configuration where role = 'p' and content > -1;
+-- The first resource group has 102 MB of memory available. 6 slices consume
```
How was 102 calculated?

KnightMurloc replied:
explained

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,105 @@
+-- 12 * 6 = 72 MB, plus one slice explicitly allocates 40 MB of memory.
+-- It is expected that the query will also be canceled due to lack of memory in
+-- the resource group and not in the Vmem tracker.
+1&: select * from alloc_view a1
```
Can we use WITH instead of views?

KnightMurloc replied:
done

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,105 @@
+1: set role test1;
+1&: select count(*) from t1 where pg_sleep(1) is not null;
+
+select segment, mem.* from gp_toolkit.gp_resgroup_status, json_object_keys(memory_usage)
```
can we execute the query as test2?

KnightMurloc replied:
done

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,105 @@
+alter resource group default_group set memory_spill_ratio 10;
+
+alter resource group default_group set memory_limit 10;
+alter resource group default_group set memory_spill_ratio 10;
```
we change default_group settings second time

KnightMurloc replied:
fixed

KnightMurloc committed:
rework test

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/92267

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,109 @@
+select gp_inject_fault('executor_pre_tuple_processed', 'suspend', dbid)
+    from gp_segment_configuration where role = 'p' and content > -1;
+1&: select count(*) from t1;
+select gp_wait_until_triggered_fault('executor_pre_tuple_processed', 1, dbid)
+    from gp_segment_configuration where role = 'p' and content > -1;
```
What happens without this query?

KnightMurloc replied:
Here we are waiting for all backends (slices) to begin up. This is necessary for the stability of the test.

andr-sokolov commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,109 @@
+select gp_inject_fault_infinite('executor_pre_tuple_processed', 'suspend', dbid)
+    from gp_segment_configuration where role = 'p' and content > -1;
+-- The first resource group has 682 * 0.15 = 102 MB of memory available.
+-- 6 slices consume 12 * 6 = 72 MB, plus one slice explicitly allocates 40 MB of
```
Can we ask OS how much memory a slice consumes?

KnightMurloc replied:
It seems that there is no known and well-documented way to get the exact amount of private memory of a process. Therefore, this is a topic for a separate study.

KnightMurloc committed:
add function removal

KnightMurloc committed:
Merge branch 'adb-6.x-dev' into ADBDEV-6156

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/92334

bandetto commented on src/test/regress/regress_gp.c:
```diff
@@ -641,8 +643,26 @@ resGroupPalloc(PG_FUNCTION_ARGS)
 	size = ceilf(memLimit * ratio);
+	// At startup, the backend process is already consuming some amount of
+	// memory. In order not to complicate the logic of the tests, we take this
+	// memory into account when allocating memory for tests.
```
```suggestion
	/* 
	 * At startup, the backend process is already consuming some amount of
	 * memory. In order not to complicate the logic of the tests, we take this
	 * memory into account when allocating memory for tests.
	 */
```

KnightMurloc replied:
done

KnightMurloc committed:
update comment

KnightMurloc committed:
Merge branch 'adb-6.x-dev' into ADBDEV-6156

bandetto commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,113 @@
+-- Allocate all the memory in the second resource group.
+select hold_memory_by_percent_on_qe(1);
+-- Wait for all backends to run before executing the query. This is necessary
+-- to avoid executing hold_memory before all backends are up.
```
What is `hold_memory`?

bandetto commented on src/test/isolation2/input/resgroup/resgroup_startup_memory.source:
```diff
@@ -0,0 +1,113 @@
+-- The first resource group has 682 * 0.15 = 102 MB of memory available.
+-- 6 slices consume 12 * 6 = 72 MB, plus one slice explicitly allocates 40 MB of
+-- memory, 72 + 40 = 112 MB. It is expected that the query will be canceled due
+-- to lack of memory in the resource group but not in the Vmem tracker.
```
> It is expected that the query will be canceled due to lack of memory in the resource group but not in the Vmem tracker.

What is the expected behavior without this patch? That the Vmem Tracker will cancel the query instead?

KnightMurloc replied:
yes

KnightMurloc committed:
fix comment in test
